modules/collect/src/main/java/com/opengamma/strata/collect/MapStream.java

```diff
@@ -282,6 +282,47 @@ private MapStream(Stream<Map.Entry<K, V>> underlying) {
     return underlying.map(e -> mapper.apply(e.getKey(), e.getValue()));
   }
 
+  //-------------------------------------------------------------------------
+  /**
+   * Transforms the entries in the stream by applying a mapper function to each key and value to produce a stream of
+   * entries, and then flattening the resulting stream of streams.
+   *
+   * @param mapper  a mapper function whose return values are included in the new stream
+   * @param <nK>  the type of the new keys
+   * @param <nV>  the type of the new values
+   * @return a stream containing the values returned from the mapper function
+   */
+  public <nK, nV> MapStream<nK, nV> flatMap(BiFunction<? super K, ? super V, Stream<Map.Entry<nK, nV>>> mapper) {
+    return wrap(underlying.flatMap(e -> mapper.apply(e.getKey(), e.getValue())));
+  }
+
+  /**
+   * Transforms the keys in the stream by applying a mapper function to each key.
+   * <p>
+   * The new keys produced will be associated with the original value.
+   *
+   * @param mapper  a mapper function whose return values are the keys in the new stream
+   * @param <R>  the type of the new keys
+   * @return a stream containing the values returned from the mapper function
+   */
+  public <R> MapStream<R, V> flatMapKeys(Function<? super K, Stream<R>> mapper) {
+    return wrap(underlying.flatMap(e -> mapper.apply(e.getKey()).map(newKey -> entry(newKey, e.getValue()))));
+  }
+
+  /**
+   * Transforms the keys in the stream by applying a mapper function to each key.
+   * <p>
+   * The new keys produced will be associated with the original value.
+   *
+   * @param mapper  a mapper function whose return values are the keys in the new stream
+   * @param <R>  the type of the new keys
+   * @return a stream containing the values returned from the mapper function
+   */
+  public <R> MapStream<R, V> flatMapKeys(BiFunction<? super K, ? super V, Stream<R>> mapper) {
+    return wrap(underlying
+        .flatMap(e -> mapper.apply(e.getKey(), e.getValue()).map(newKey -> entry(newKey, e.getValue()))));
+  }
+
```

Comment:
`flatMapValues()` does make sense after all, so those 2 methods will need adding.

Reply:
```
  /**
   * Transforms the values in the stream by applying a mapper function to each value.
   * <p>
   * The new values produced will be associated with the original key.
   * <p>
   * For example this could turn a {@code MapStream<LocalDate, List<String>>} into a
   * {@code MapStream<LocalDate, String>}
   *
   * @param mapper  a mapper function whose return values are the keys in the new stream
   * @param <R>  the type of the new keys
   * @return a stream containing the values returned from the mapper function
   */
  public <R> MapStream<K, R> flatMapValues(Function<? super V, Stream<R>> mapper) {
    return wrap(underlying.flatMap(e -> mapper.apply(e.getValue()).map(newValue -> entry(e.getKey(), newValue))));
  }
```